frontend/src/components/Recipe.jsx

```diff
@@ -1,39 +1,167 @@
+import Loader from './Loader'
+import { Helmet } from 'react-helmet'
+import AddStep from './AddStep'
+import AddIngredient from './AddIngredient'
 
-import RecipeEdit from '../containers/RecipeEdit'
-import RecipeViewing from './RecipeViewing'
+import StepContainer from './StepContainer'
+import Ingredient from './Ingredient'
+import RecipeTitle from './RecipeTitle'
 
-import { inputAbs } from '../input'
+import {
+  addingRecipeIngredient,
+  addingRecipeStep,
+  fetchRecipe,
+  deletingRecipe,
+  updateRecipe
+} from '../store/actions'
 
-class Recipe extends React.Component {
-  state = {
-    editing: false,
+const mapStateToProps = (state, props) => {
+  const id = props.match.params.id
+  const recipe = state.recipes[id]
+    ? state.recipes[id]
+    : { loading: true }
+  return recipe
+}
+
+const mapDispatchToProps = dispatch => ({
+  fetchRecipe: id => dispatch(fetchRecipe(id)),
+  addIngredient: (recipeID, ingredient) =>
+    dispatch(addingRecipeIngredient(recipeID, ingredient)),
+  addStep: (id, step) =>
+    dispatch(addingRecipeStep(id, step)),
+  update: (id, data) => dispatch(updateRecipe(id, data)),
+  remove: id => dispatch(deletingRecipe(id)),
+})
+
+@connect(
+  mapStateToProps,
+  mapDispatchToProps
+)
+export default class Recipe extends React.Component {
+  static propTypes = {
+    id: PropTypes.number.isRequired,
+    name: PropTypes.string.isRequired,
+    author: PropTypes.string.isRequired,
+    source: PropTypes.string.isRequired,
+    servings: PropTypes.string.isRequired,
+    time: PropTypes.string.isRequired,
+    ingredients: PropTypes.array.isRequired,
+    steps: PropTypes.array.isRequired,
+    loading: PropTypes.bool.isRequired,
+    error404: PropTypes.bool.isRequired,
+    owner: PropTypes.object.isRequired,
+    update: PropTypes.func.isRequired,
+    remove: PropTypes.func.isRequired,
+    deleting: PropTypes.bool.isRequired,
+  }
+
+  // necessary as these are undefined between page load and data fetch
+  static defaultProps = {
+    id: 0,
+    name: '',
+    author: '',
+    source: '',
+    servings: '',
+    time: '',
+    deleting: false,
+    removing: false,
+    ingredients: [],
+    steps: [],
+    loading: true,
+    error404: false,
+    owner: {
+      type: 'user',
+      id: 0,
+      name: '',
+    },
+    updating: false,
   }
 
-  handleInputChange = e =>
-    this.setState({ [e.target.name]: inputAbs(e.target.value) })
+  state = {
+    show: false,
+    addStep: false,
+    addIngredient: false,
+  }
 
-  componentWillMount = () => {
+  componentWillMount () {
     this.props.fetchRecipe(this.props.match.params.id)
   }
 
   render () {
-    if (this.state.editing) {
-      return (
-        <RecipeEdit
-          { ...this.props }
-          cancelEdit={ () => this.setState({ editing: false }) }
-        />
-      )
+    const {
+      id,
+      name,
+      ingredients,
+      steps,
+      loading,
+      error404,
+    } = this.props
+    if (error404) {
+      return <NoMatch/>
```

Comment:
Can this dispatch a redux action to go to the 404 page. It should be the option that replaces the current url.